### PR TITLE
Implement labelling

### LINKED ISSUE #64: [FEATURE] Create utility function to apply {labelled} labels to supertibble
# Feature Request Description
A new exported function (`make_labelled()`) that uses the [`labelled`](http://larmarange.github.io/labelled/) package to apply labels to fields in the supertibble and fields within the supertibble's list columns.

# Proposed Solution
A use would run:
```r
suppertibble <- make_labelled(suppertibble)
```
which would:

1. Apply labels to the variables in the supertibble. These will be defined by `REDCapTidieR`.
2. Apply labels to the variables in `supertibble$redcap_metadata` and `supertibble$redcap_events`. These will be defined by `REDCapTidieR`.
3. Apply labels to the variables in `supertibble$redcap_data`. These would be derived from the contents of `supertibble$redcap_metadata`

Additionally we should add functionality for editing the labels in `redcap_metadata` before applying them to `redcap_data`. Two potential options:
1. Provide additional parameters to `make_labelled()` that control custom formatting, ex. `remove_terminal_colons`
2. Add a parameter that let's the user pass a formatting function to preprocess labels and give a sensible default, ex:
```r
formatter <- function(x) stringr::str_remove(x, ":$")

suppertibble <- make_labelled(suppertibble, label_format = formatter)
```

Finally, rather than importing `labelled` we should check that it's installed in `make_labelled()` and issue an error message if it isn't that asks the user to install it.

### Checklist

- [x] The issue is atomic
- [x] The issue description is documented
- [x] The issue title describes the problem succinctly
- [x] Developers are assigned to the issue
- [x] Labels are assigned to the issue

### PR DESCRIPTION
**Pull request by ezraporter**
# Description
This PR implements the `make_labelled()` functionality described in #64. At this stage the core functionality with test cases 
 and reasonable error messages is complete. The remaining additions to complete the issue are:

- Implement custom formatter parameter in `make_labelled()`
- Add an exported default formatter
- Flesh out documentation
- Update vignette to show custom labelling + custom formatting functionality

# Proposed Changes 

Add exported `make_labelled()` function that:
- Applies predefined labels to columns of supertibble
- Applies labels to columns of elements of `supertibble$redcap_metadata` by replacing underscores with spaces and converting to title case
    - ex. `field_name` -> `"Field Name"`
    - I'm not set on this solution but it seemed to do the trick pretty well rather than predefining labels for all the metadata fields
-  Applies labels to columns of elements of `supertibble$redcap_data` based on the contents of `supertibble$redcap_metadata`

Input Checking:
- Input must be a `data.frame`/`tibble`
    - using `checkmate` for this
- Input must contain `redcap_data` and `redcap_metadata` columns
    - See error screenshot 1 
- Every element of `redcap_metadata` must have `field_name` and `field_label` columns
    - See error screenshot 2
- `labelled` must be installed
    - I'm using `rlang::check_installed()` for this. It does the check and also offers to install it for the user if it's missing. 

**Screenshots**
 Error 1:
Screen Shot 2022-11-03 at 1 22 23 PM

Error 2:
Screen Shot 2022-11-03 at 1 22 58 PM

<details>
  <summary>Code to produce errors</summary>

```r
devtools::load_all()

data <- tibble::tribble(
  ~ redcap_form_name, ~ redcap_data, ~ redcap_metadata,
  "form 1", tibble(x = letters[1:3]), tibble(field_name = "x", field_label = "X Label"),
  "form 2", tibble(y = letters[1:3]), tibble(field_name = "y", field_label = "Y Label")
)

data |>
  dplyr::select(-redcap_data) |>
  make_labelled()

data$redcap_metadata[[1]]$field_label <- NULL

make_labelled(data)
```
</details>

### Issue Addressed
Closes #64

### PR Checklist
Before submitting this PR, please check and verify below that the submission meets the below criteria:

- [x] New/revised functions have associated tests
- [x] New tests that make API calls use `httptest::with_mock_api` and any new mocks were added to `tests/testthat/fixtures/create_httptest_mocks.R`
- [x] New/revised functions use appropriate naming conventions
- [x] New/revised functions don't repeat code
- [x] Code changes are less than **250** lines total
- [x] Issues linked to the PR using [GitHub's list of keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [ ] The appropriate reviewer is assigned to the PR
- [x] The appropriate developers are assigned to the PR

# Code Review
This section to be used by the reviewer and developers during Code Review after PR submission

### Code Review Checklist

- [ ] I checked that new files follow naming conventions and are in the right place
- [ ] I checked that documentation is complete, clear, and without typos
- [ ] I added/edited comments to explain "why" not "how"
- [ ] I checked that all new variable and function names follow naming conventions
- [ ] I checked that new tests have been written for key business logic and/or bugs that this PR fixes
- [ ] I checked that new tests address important edge cases

### CODE REVIEW

**Review by rsh52 — COMMENTED** [review]
Admittedly need to do a deeper dive into the actual functionality and test writing, but wanted to do a quick review and get you some feedback before you're off tomorrow. 

One clarification question on the usefulness of the `check_req_labelled_metadata_fields` function. Also I think you're testing for these under `make_labelled()` tests for errors, but we do have a dedicated test location for `check`s. Might be good to have there instead/in addition.

I tested using `make_labelled` on the Super Heroes database and the output looks exactly like what we want, so I'm happy saying the code works! (`formatter` will take care of the ":" once that's implemented). Promise to do a deeper dive tomorrow, but currently I think this looks well-written 👍

**Comment by skadauke** [comment]
This is a great start! I looked at the Prodigy database and found a couple of minor issues. Some of them should probably be made into separate GitHub issues.

- Multi checkboxes need special handling to show both the field label and also the relevant option.
- In the supertibble, the `na_pct` field appears to show the fraction of NA. Multiply by 100?
- Some field labels include HTML tags because our Prodigy database is so fancy! I think we should have a formatter `strip_html_tags()`. 
- Another formatter that would useful is one that removes terminal periods
- “Other” fields that are embedded next to the “Other” answer choice of a radio/dropdown should have the field embedding and any whitespace before the opening curly bracket removed from the response. If this doesn't make any sense, see: https://youtu.be/izviJ81GJoI?t=336. E.g.: igg_laboratory: “Other {igg_laboratory_specify}” should be “Other”

Can I also request some changes in the static labels:

- Supertibble labels
  - REDCap Instrument Name
  - REDCap Instrument Description
  - Data
  - Metadata
  - Events and Arms Associated with this Instrument
  - Repeating or Nonrepeating?
  - # of Rows in Data
  - # of Columns in Data
  - Data size in Memory
  - % of Data Missing
- Data labels – make the standard REDCap fields stand out a bit more
  - REDCap Repeat Instance
  - REDCap Event
  - REDCap Arm
  - REDCap Form Status
- Metadata labels – we might want to use the same ones that you will get when downloading Data Dictionary as Excel sheet from REDCap directly
  - Variable / Field Name
  - Field Label	
  - Field Type
  - Section Header
  - Form Name
  - Field Note
  - Text Validation Type OR Show Slider Number
  - Text Validation Min
  - Text Validation Max
  - Identifier?
  - Branching Logic (Show field only if...)
  - Required Field?
  - Custom Alignment
  - Question Number (surveys only)
  - Matrix Group Name
  - Matrix Ranking?
  - Field Annotation

**Comment by skadauke** [comment]
Another thought I had for a formatter is the following. We've been using the Field Note to specify the unit for numeric fields. We haven't used the Field Note for anything other than specifying units. I'm wondering if we could have a formatter that checks if there's something in Field Note and if so, adds it in parentheses after the field.

Example--

Field label: `Weight`
Field note: `kg`
Formatter turns this into: `Weight (kg)`

**Comment by ezraporter** [comment]
Pushed changes to take care of all the non-formatter tweaks:
1. Moved test code for checks into `test-checks`
2. Adopted @skadauke's label suggestions
3. Added labels for `redcap_events` tibbles

@rsh52, I know we just discussed the opposite but are you good with including the change to `na_pct` in a different PR? That has the benefit of keeping this PR from touching `read_redcap_tidy()`. I can make a new issue for it if it doesn't fit in the issue you were going to make.

**Comment by ezraporter** [comment]
> * Event Description (I know we don't have this yet)

Good reminder that I should add an issue for this. I had forgotten about it

**Comment by ezraporter** [comment]
> That's fine, I would put it in a separate issue though (maybe we can tackle them both in one PR). It doesn't really fit into the issue I wrote now.

Sounds good. I'll make it

**Comment by ezraporter** [comment]
> * Multi checkboxes need special handling to show both the field label and also the relevant option. Example:
> 
> | Field Name                                      | Desired Label                               |
> |-------------------------------------------------|---------------------------------------------|
> | crs_anti_tcell_therapy_types___cyclophosphamide | Anti-T Cell therapy given: Cyclophosphamide |
> | crs_anti_tcell_therapy_types___atg              | Anti-T Cell therapy given: ATG              |
> | crs_anti_tcell_therapy_types___other            | Anti-T Cell therapy given: Other            |

This is proving to be more difficult than I had hoped because we currently **don't** return the checkbox option label in the metadata. Ex. All we currently get is `crs_anti_tcell_therapy_types___cyclophosphamide`, not the option label "Cyclophosphamide". Parsing that string won't even work in general because it uses the raw value and not the label.

## Proposed Solution

I think the best solution is actually not to have any special handling for checkbox fields in `make_labelled()`. Instead I think we should change how we handle the metadata in `read_redcap_tidy()`. Currently `update_field_names()` expands out the metadata to include a row for each checkbox option but in doing so just carries over the `field_label`.

Input:
|field_name|field_label|select_choices_or_calculations|
|----------------------|----------------------|-------------------------|
crs_anti_tcell_therapy_types | Anti-T Cell therapy given: | cyclophosphamide, Cyclophosphamide \| atg, ATG \| other, Other {crs_anti_tcell_therapy_types_specify}|

Current Output:
|field_name|field_label|updated_field_name|
|----------------------|----------------------|-------------------------|
crs_anti_tcell_therapy_types | Anti-T Cell therapy given: | crs_anti_tcell_therapy_types___cyclophosphamide |
crs_anti_tcell_therapy_types | Anti-T Cell therapy given: | crs_anti_tcell_therapy_types___atg |
crs_anti_tcell_therapy_types | Anti-T Cell therapy given: | crs_anti_tcell_therapy_types___other |

My suggestion is to change `update_field_names()` so that the labels get added to `field_label`:

New Output:
|field_name|field_label|updated_field_name|
|----------------------|----------------------|-------------------------|
crs_anti_tcell_therapy_types | Anti-T Cell therapy given: Cyclophosphamide | crs_anti_tcell_therapy_types___cyclophosphamide |
crs_anti_tcell_therapy_types | Anti-T Cell therapy given: ATG | crs_anti_tcell_therapy_types___atg |
crs_anti_tcell_therapy_types | Anti-T Cell therapy given: Other | crs_anti_tcell_therapy_types___other |

If we do this everything will flow into the `make_labelled()` code correctly.

**Comment by ezraporter** [comment]
Added an initial implementation of the formatting functionality. Looking for feedback on that before refining it. The basic usage is:

```r
supertbl <- tibble::tribble(
    ~ redcap_data, ~ redcap_metadata,
    tibble(x = letters[1:3]), tibble(field_name = "x", field_label = "<b>X Label:</b>")
  )

# Use the labels as is (default behavior):

make_labelled(supertbl)
make_labelled(supertbl, format_labels = NULL)

# Apply default formatting:

## using the exported function
make_labelled(supertbl, format_labels = format_labels_default)
## or referencing it by name
make_labelled(supertbl, format_labels = "default")

# Supply your own formatting function

make_labelled(supertbl, format_labels = tolower)
```

File: R/labelled.R [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+  if (is.null(format_labels)) {
+    formatter <- identity
+  } else if (is.function(format_labels)) {
+    formatter <- format_labels
+  } else if (format_labels == "default") {
+    formatter <- format_labels_default
+  } else {
+    # TODO: add informative error with cli
+    stop()
```
Comment (rsh52) [inline R/labelled.R:56]: To immediate mind we'd want a catch like:

```
"✖ ERROR: argument `formatter` must be of type 'function'" 
ℹ "Or use "default" for `REDCapTidieR` default formatting options".
```
Or something better worded to point users to default settings. I can't think of much else we would check for here, if a custom formatter fails it's mostly the user's responsibility.

File: R/labelled.R [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+  )
+
+  # Apply labels ----
+
+  # Utility function for label setting
+  # Set labels of tibble from named vector but don't fail if labels vector has
+  # variables that aren't in the data
+  safe_set_variable_labels <- function(data, labs) {
+    labs_to_keep <- intersect(names(labs), colnames(data))
+    labelled::set_variable_labels(data, !!!labs[labs_to_keep])
```
Comment (rsh52) [inline R/labelled.R:117]: Unsure if it's required to add an `@importFrom` since you're calling out the package here, but to keep consistency I'd throw `@importFrom laballed set_variable_labels` into the roxygen docs/`NAMESPACE`.

**Review by rsh52 — COMMENTED** [review]
Couple of comments, but otherwise I think this is getting close to ready for merge.

File: R/labelled.R (hunk 2 above)
Comment (ezraporter) [reply R/labelled.R:117]: When I do this and run `check()` I get:

```
checking package dependencies ... ERROR
Namespace dependency missing from DESCRIPTION Imports/Depends entries: ‘labelled’
```

Unless I'm missing something. I don't think you can have things in the `NAMESPACE` that aren't in `Imports/Depends` and `@importFrom` adds to the `NAMESPACE`.

File: R/labelled.R (hunk 1 above)
Comment (ezraporter) [reply R/labelled.R:56]: Yeah something like this makes sense to me. I'll add it when I sure up the interface and documentation

File: R/checks.R [hunk 3]
```diff
@@ -173,3 +173,101 @@ check_forms_exist <- function(db_metadata, forms) {
...
+#'
+#' @param supertbl a supertibble containing a \code{redcap_metadata} column
+#'
+#' @return
+#' an error message alerting that instrument metadata is incomplete
+#'
+#' @keywords internal
+check_req_labelled_metadata_fields <- function(supertbl) {
+
+  req_fields <- c("field_name", "field_label")
```
[the remark this answers is not in the document]
Comment (ezraporter) [reply R/checks.R:226]: I don't remember where we landed on this. Did we say keep this check or keep things simple and cut it? Cutting it had the benefit of stripping out some slightly hacky error-message-generation code on my part.

File: R/labelled.R (hunk 2 above)
Comment (rsh52) [reply R/labelled.R:117]: Right, in this case you would need to have it in the `DESCRIPTION` file via `use_package(__)`. We're currently doing that with `cli` which isn't a user-side thing, so I figure it doesn't hurt to do it for `labelled` as well, but I'm not really that opinionated one way or another.

We may also not need to do this anyway since you're checking for `labelled` installation in the new function anyway.

Comment (ezraporter) [reply R/labelled.R:117]: I think the pattern we were envisioning is that the user only needs to have `labelled` installed if they want to use `make_labelled()`. So we put it in `Suggests` and check that the user has it installed at the top of `make_labelled()`.

The problem is that to use `@importFrom` we'd need to bump it from `Suggests` to `Imports` which would mean that it gets installed when a user installs `REDCapTidieR`.

Comment (rsh52) [reply R/labelled.R:117]: Ah, ok makes sense to me 👍

File: R/checks.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (rsh52) [reply R/checks.R:226]: I think we agreed it's unlikely it'll ever happen but there's an off-chance a user could manipulate these variable names. The processing time was negligible so it wasn't hurting anything there.

**Comment by skadauke** [comment]
> Another thought I had for a formatter is the following. We've been using the Field Note to specify the unit for numeric fields. We haven't used the Field Note for anything other than specifying units. I'm wondering if we could have a formatter that checks if there's something in Field Note and if so, adds it in parentheses after the field.
> 
> Example--
> 
> Field label: `Weight` Field note: `kg` Formatter turns this into: `Weight (kg)`

More thoughts on this... it could be argued that a label formatter function should have access not only to the `field_label` but to _all_ the information in the metadata for a specific field. Maybe formatter functions should have a second optional argument `metadata` that contains a list of the fields of the metadata for the specific field into the function. You could then do something like this:

``` r
fmt_append_units_from_field_note <- function(x, metadata) {
  x %>%
    glue::glue("{x} ({metadata$field_note})")
}
```

**Comment by ezraporter** [comment]
> > Another thought I had for a formatter is the following. We've been using the Field Note to specify the unit for numeric fields. We haven't used the Field Note for anything other than specifying units. I'm wondering if we could have a formatter that checks if there's something in Field Note and if so, adds it in parentheses after the field.
> > Example--
> > Field label: `Weight` Field note: `kg` Formatter turns this into: `Weight (kg)`
> 
> More thoughts on this... it could be argued that a label formatter function should have access not only to the `field_label` but to _all_ the information in the metadata for a specific field. Maybe formatter functions should have a second optional argument `metadata` that contains a list of the fields of the metadata for the specific field into the function. You could then do something like this:

I need to think a little bit more about the implications of this while I mess around with the formatter code tomorrow.

The worry I have is that currently the formatter treats every field the same. I'm not sure how useful opening up all the metadata is without giving the user field-level control. Ex. how useful is appending the field note if you **have to** do it for every field in the project? I know we have redcaps like this now but what happens when a stakeholder says "hey, we really want this informative field note next to this field"?

At the same time, I'm having trouble imagining what the API would like if we wanted to give users field-level control over formatting 🤔

**Comment by skadauke** [comment]
That's a good point. Maybe the concerns of applying universal rules for formatting and applying field-level fixes for formatting could be separated? Like:

``` r
make_labelled(...) %>%
fix_labels(
   some_field = "{field_label} ({field_note}),
   some_other_field = "This is what I want this field label to be",
   ...
)

?
________________________________
From: ezraporter ***@***.***>
Sent: Thursday, November 10, 2022 5:14 PM
To: CHOP-CGTDataOps/REDCapTidieR ***@***.***>
Cc: Stephan Kadauke ***@***.***>; Mention ***@***.***>
Subject: [External]Re: [CHOP-CGTDataOps/REDCapTidieR] Implement labelling (PR #65)


Another thought I had for a formatter is the following. We've been using the Field Note to specify the unit for numeric fields. We haven't used the Field Note for anything other than specifying units. I'm wondering if we could have a formatter that checks if there's something in Field Note and if so, adds it in parentheses after the field.
Example--
Field label: Weight Field note: kg Formatter turns this into: Weight (kg)

More thoughts on this... it could be argued that a label formatter function should have access not only to the field_label but to all the information in the metadata for a specific field. Maybe formatter functions should have a second optional argument metadata that contains a list of the fields of the metadata for the specific field into the function. You could then do something like this:

I need to think a little bit more about the implications of this while I mess around with the formatter code tomorrow.

The worry I have is that currently the formatter treats every field the same. I'm not sure how useful opening up all the metadata is without giving the user field-level control. Ex. how useful is appending the field note if you have to do it for every field in the project? I know we have redcaps like this now but what happens when a stakeholder says "hey, we really want this informative field note next to this field"?

At the same time, I'm having trouble imagining what the API would like if we wanted to give users field-level control over formatting 🤔

—
Reply to this email directly, view it on GitHub<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCHOP-CGTDataOps%2FREDCapTidieR%2Fpull%2F65%23issuecomment-1310971760&data=05%7C01%7Ckadaukes%40chop.edu%7Cdec53caece7d459d215308dac368fa48%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638037152901724489%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=kfeML3gHsslgPRnKRhPFFCKj1Q8AUKQaJjX2Llsbk8c%3D&reserved=0>, or unsubscribe<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACTGHWSKYXNXTKZN2XTYCHTWHVXVJANCNFSM6AAAAAARWMDGHM&data=05%7C01%7Ckadaukes%40chop.edu%7Cdec53caece7d459d215308dac368fa48%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638037152901724489%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=x5dwU%2BSwgeCaDnNBUl3WBMiYcu7jAAgy%2BR%2FCIg3Mpy0%3D&reserved=0>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

** This email originated from an EXTERNAL sender to CHOP. Proceed with caution when replying, opening attachments, or clicking links. Do not disclose your CHOP credentials, employee information, or protected health information to a potential hacker**.

**Comment by ezraporter** [comment]
Closing this PR to rebase and open a clean one with final updates
